### PR TITLE
Bug 2066196: ensure secrets have backup labels

### PR DESCRIPTION
**Pull request by vrutkovs**
ACM 2.5 will GA the DR (disaster recovery) feature, which will automatically backup all required resources created by each component on hub side, then in case the hub goes down, DR can use the backed up resources to recreate the info on a new hub. In order for AI ClusterDeployments to support this all generated secrets need to have either `hive.openshift.io/secret-type` (kubeconfig) or `cluster.open-cluster-management.io/backup` labels (agent auth, postgres secret, BMH secret).

`kubeconfig` is already labelled with `hive.openshift.io/secret-type=kubeconfig`, so this PR labels remaining secrets with  `cluster.open-cluster-management.io/backup=true`

TODO:
* [x] Break out label part from `getSecret` into `addLabeltoSecret` to keep it cleaner

## List all the issues related to this PR

- [x] Bug fix

## What environments does this code impact?

- [x] Operator Managed Deployments

## How was this code tested?

- [x] Manual (Elaborate on how it was tested)

Not entirely sure how to test that the whole env can be backed up / restored

## Assignees

/cc @avishayt 
/cc @omertuc

### CODE REVIEW

**Comment by omertuc** [comment]
What is:

>restore Hive state on different server 

Can you elaborate?

File: internal/controller/controllers/common.go [hunk 1]
```diff
@@ -64,6 +64,15 @@ func getSecret(ctx context.Context, c client.Client, r client.Reader, key types.
 			return nil, errors.Wrapf(err, errorMessage)
 		}
 	}
+
+	// Add backup label to the secret if not present
+	metav1.SetMetaDataLabel(&secret.ObjectMeta, backupLabel, "true")
+	err := c.Update(ctx, secret)
+	if err != nil {
+		errorMessage = fmt.Sprintf("failed to set label %s:%s for secret %s/%s", backupLabel, WatchResourceValue, key.Namespace, key.Name)
```
Comment (omertuc) [inline internal/controller/controllers/common.go:72]: `WatchResourceValue` is irrelevant

**Comment by vrutkovs** [comment]
>Can you elaborate?

Updated descritpion - ACM Dr is a 2.5 feature allowing ClusterDeployments to backup data and restore state on a different cluster

**Comment by omertuc** [comment]
Could you update the commit message as well? Also it's not new in this PR but it's a bit unfortunate that `getSecret` updates the secret with additional values...

From a PR point of view lgtm but from a design point of view I'm really not sure what's going on here - why are data backups not including all secrets in the first place? Why require all these special labels?

The obvious answer is that this including those secrets in the backup is problematic from an ACM security standpoint, so why should we ignore those security considerations completely when creating / processing (`getSecret`) AI-related secrets? We're basically opting the user into giving up their kubeconfig / kubeadmin credentials into hub cluster backups, something that is not immediately obvious to me whether is the correct thing to do

**Comment by vrutkovs** [comment]
/hold

for manual testing


>It's not new in this PR but it's a bit unfortunate that getSecret updates the secret with additional values... that's a bit unexpected, given the function's name.

Good point, added a TODO

>why are data backups not including all secrets in the first place?

The idea is to restore backups on a different server. These certainly would include sensitive data

**Comment by filanov** [comment]
you will need to backport so please open a BZ for this issue

**Comment by vrutkovs** [comment]
IIUC we don't need to mark CRs (these are backed up automatically), just the secrets we generate on the way (as it can't save things like SA secrets etc.)

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 2]
```diff
@@ -84,6 +84,9 @@ const (
 	imageServiceStatefulSetFinalizerName = imageServiceName + "." + aiv1beta1.Group + "/ai-deprovision"
 	agentServiceConfigFinalizerName      = "agentserviceconfig." + aiv1beta1.Group + "/ai-deprovision"
 
+	backupLabel      = "cluster.open-cluster-management.io/backup"
```
[the remark this answers is not in the document]
Comment (vrutkovs) [reply internal/controller/controllers/agentserviceconfig_controller.go]: Good idea, this probably belongs there better

**Comment by filanov** [comment]
> IIUC we don't need to mark CRs (these are backed up automatically), just the secrets we generate on the way (as it can't save things like SA secrets etc.)

Does it apply to https://github.com/openshift/assisted-service/blob/608ee245463dcc15b131af2059853bcd11b3ef7d/internal/controller/controllers/clusterdeployments_controller.go#L540 ?

**Comment by filanov** [comment]
@birsanv related to https://github.com/stolostron/backlog/issues/18513#issuecomment-1064317512 
Which CRs are you handling and what do we need to handle?

**Comment by vrutkovs** [comment]
>Does it apply to createClusterCredentialSecret

Yes, its already being marked with `hive.openshift.io/secret-type`

**Comment by romfreiman** [comment]
Are we certain about not adding labels to all crs
https://github.com/stolostron/backlog/issues/18513#issuecomment-1064317512

Seems that they expect them to be labeled explicitly, no?

**Comment by vrutkovs** [comment]
>Are we certain about not adding labels to all crs

I didn't get a definitive answer on that (also don't have access to backlog issue for some reason).

In any case, I'd prefer this PR to focus on secrets. CRs probably should be handled in a different issue - or maybe we want to clarify the JIRA issue (it clearly states "secrets")

**Comment by vrutkovs** [comment]
/hold cancel
/skip
CAPI job is permafailing in periodics too

**Review by filanov — APPROVED** [review]

**Comment by filanov** [comment]
/hold 
Until manual tests are completed.

**Pull request merged**